### PR TITLE
add option to send Server-Timing header when replying CONNECT request

### PR DESCRIPTION
**Pull request by joesis**
For https://github.com/getlantern/lantern-internal/issues/2151

With this header, the client can have a more accurate RTT measurement across all transports. Unlike connection establishment, CONNECT request and response is small enough to fit into a single IP packet, and happens more frequently for multiplexed transports, so we can eliminate many of the periodical checks and hacks to deal with lampshade. The plan is to use this and the RTT of 1kb http ping (which fits into a single packet too) to measure latency.

### CODE REVIEW

**Comment by myleshorton** [comment]
I generally like this @joesis, although AFAICT the current code largely measures the upstream connection to the destination -- will comment in the code itself.

File: proxy_connect.go [hunk 1]
```diff
@@ -93,6 +98,11 @@ func (proxy *proxy) nextCONNECT(downstream net.Conn) filters.Next {
 		// for example, if some proxy servers reside in jurisdictions where an
 		// origin site is blocked but other proxy servers don't.
 		resp, nextCtx = respondOK(resp, modifiedReq, nextCtx)
+		if proxy.OKSendsServerTiming {
+			millis := fmt.Sprintf("dialupstream;dur=%d", time.Since(start)/time.Millisecond)
+			resp.Header.Add(serverTimingHeader, millis)
+		}
+
```
Comment (myleshorton) [inline proxy_connect.go]: Yeah it looks like this is primarily measuring the upstream connect and response times right @joesis? I wonder if we could tap into the data in the tracing context if/when we add that, or somehow sync time measurements between client and server...

Comment (myleshorton) [reply proxy_connect.go]: Oh I see sorry I misunderstood -- this would allow the client to subtract the upstream time right? Cool I like it...

**Comment by myleshorton** [comment]
I like the idea of adding travis or some CI here as well. Maybe just get that fully working, and then I can merge? I definitely agree this approach is solid since it lets us take advantage of traffic we're constantly sending.

**Comment by myleshorton** [comment]
actually I'm just gonna pull travis for now and shoot that to a separate branch!!

**Comment by myleshorton** [comment]
OK just pushed the ci stuff to the add-ci branch @joesis -- would be awesome to add I agree!

**Comment by myleshorton** [comment]
This is deployed everywhere @joesis

**Comment by joesis** [comment]
Thanks @myleshorton and yeah the travis thing should be in a separate branch.

**Pull request merged**